File: src/python/challenges/problems/unique_paths_ii_test.py

# https://leetcode.com/problems/unique-paths-ii

from typing import List
from core.problem_base import *

class UniquePathsIi(ProblemBase):
    def Solution(self, obstacleGrid: List[List[int]]) -> int:
        width = len(obstacleGrid)
        height = len(obstacleGrid[0])

        for row in range(width):
            for col in range(height):
                if obstacleGrid[row][col] == 1:
                    obstacleGrid[row][col] = 0
                    continue

                if row == 0 and col == 0:
                    obstacleGrid[row][col] = 1
                    continue

                if row - 1 >= 0:
                    obstacleGrid[row][col] += obstacleGrid[row - 1][col]

                if col - 1 >= 0:
                    obstacleGrid[row][col] += obstacleGrid[row][col - 1]

        return obstacleGrid[width - 1][height - 1]

if __name__ == '__main__':
    TestGen(UniquePathsIi) \
        .Add(lambda tc: tc.Param([[0,0,0,0],[0,1,0,0],[0,0,1,0],[1,1,0,0],[0,0,0,0],[0,0,0,0]]).Result(2) ) \
        .Add(lambda tc: tc.Param([[0,0,0,0],[0,1,0,0],[0,0,0,0],[0,0,0,0]]).Result(8) ) \
        .Add(lambda tc: tc.Param([[0,1],[0,0]]).Result(1) ) \
        .Add(lambda tc: tc.Param([[0,0,0],[0,1,0],[0,0,0]]).Result(2) ) \
        .Run()

File: src/go/challenges/problems/unique_paths_ii_test.go
// https://leetcode.com/problems/unique-paths-ii

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type UniquePathsIi struct{}

func TestUniquePathsIi(t *testing.T) {
	gen := core.TestSuite[UniquePathsIi]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([][]int{{0, 0, 0}, {0, 1, 0}, {0, 0, 0}}).Result(2)
	}).Add(func(tc *core.TestCase) {
		tc.Param([][]int{{0, 0, 0, 0}, {0, 1, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}}).Result(8)
	}).Add(func(tc *core.TestCase) {
		tc.Param([][]int{{1}}).Result(0)
	}).Add(func(tc *core.TestCase) {
		tc.Param([][]int{{0}}).Result(1)
	}).Run(t)
}

func (UniquePathsIi) Solution(obstacleGrid [][]int) int {
	width := len(obstacleGrid)
	height := len(obstacleGrid[0])

	for row := 0; row < width; row++ {
		for col := 0; col < height; col++ {
			if obstacleGrid[row][col] == 1 {
				obstacleGrid[row][col] = 0
				continue
			}

			if row == 0 && col == 0 {
				obstacleGrid[row][col] = 1
				continue
			}

			if row-1 >= 0 {
				obstacleGrid[row][col] += obstacleGrid[row-1][col]
			}

			if col-1 >= 0 {
				obstacleGrid[row][col] += obstacleGrid[row][col-1]
			}
		}
	}

	return obstacleGrid[width-1][height-1]
}
